libs/custom-scanner/src/custom_a4_scanner.ts

```diff
@@ -0,0 +1,632 @@
+    return this.channelMutex.withLock((channel) =>
+      getStatusInternal(channel, ONLY_VALID_JOB_ID)
+    );
+  }
+
+  /**
+   * Creates a job on the scanner. Does not hold the public lock.
+   */
+  private async createJobInternal(): Promise<Result<number, ErrorCode>> {
+    debug('creating job');
+    this.currentJobId = ONLY_VALID_JOB_ID;
+    let createJobResult!: Result<number, ErrorCode>;
+
+    for (
+      let recreateJobCount = 0, errorCount = 0;
+      recreateJobCount <= RECREATE_JOB_MAX &&
+      errorCount <= CREATE_JOB_ERROR_MAX;
+
+    ) {
+      debug('creating job (attempt %d)', recreateJobCount + errorCount + 1);
+      createJobResult = await this.channelMutex.withLock((channel) =>
+        createJob(channel)
+      );
+
+      /* istanbul ignore next */
+      if (createJobResult.isErr()) {
+        const errorCode = createJobResult.err();
+        debug('create job error: %o', errorCode);
+        if (errorCode === ErrorCode.JobNotValid) {
+          void (await this.channelMutex.withLock((channel) =>
+            endJob(channel, ONLY_VALID_JOB_ID)
+          ));
+          recreateJobCount += 1;
+        } else {
+          errorCount += 1;
+        }
+      } else {
+        break;
+      }
+    }
+
+    debug('create job result: %o', createJobResult);
+    if (createJobResult.isOk()) {
+      const jobId = createJobResult.ok();
+      debug('setting current job id to %d', jobId);
+      this.currentJobId = jobId;
+    }
+    return createJobResult;
+  }
+
+  /**
+   * Retries an attempt to send commands to the scanner up to `max` times.
+   * Fails if an error occurs that is not `ErrorCode.JobNotValid`, as this
+   * error code indicates that we should try to create a new job.
+   */
+  private async withRetries<T>(
+    fn: () => Promise<Result<T, ErrorCode>>,
+    {
+      max = 1,
+    }: {
+      max?: number;
+    } = {}
+  ): Promise<Result<T, ErrorCode>> {
+    let result!: Result<T, ErrorCode>;
+
+    for (let i = 0; i < max + 1; i += 1) {
+      debug('withRetries attempt %d/%d', i + 1, max + 1);
+
+      result = await fn();
+      if (result.isOk()) {
+        debug('withRetries success on attempt %d/%d', i + 1, max + 1);
+        break;
+      }
+
+      const error = result.err();
+      debug('withRetries error on attempt %d/%d: %o', i + 1, max + 1, error);
+      if (error !== ErrorCode.JobNotValid) {
+        break;
+      }
+    }
+
+    return result;
+  }
+
+  /**
+   * Moves a sheet of paper as directed.
+   */
+  async move(movement: FormMovement): Promise<Result<void, ErrorCode>> {
+    debug('moving %o', movement);
+    const result = await this.publicApiMutex.withLock(() =>
+      this.withRetries(async () => {
+        const createJobResult = await this.createJobInternal();
+        if (createJobResult.isErr()) {
+          return createJobResult;
+        }
+
+        return await this.channelMutex.withLock((channel) =>
+          formMove(channel, this.currentJobId, movement)
```

Comment:
As some of my other comments have said, much of this is modeled on the Custom example code. I think the scanner itself is kind of confused about what a "job" is. My hunch is that they repurposed some other code that had and really used the "job" concept, and rather than totally stripping it out they just made it be there and kinda work but you can only ever have one job and its job ID is 1. Either that or they started building job support and didn't finish and just left it halfway intact. I'm not sure.

That said, we could probably simplify our code a little bit to totally ignore the job ID they give us and instead always use the hardcoded `ONLY_VALID_JOB_ID`. :shrug:

Reply:
Ok that's helpful context. I like the idea of just using the constant (and maybe adding a comment explaining this weirdness)

Reply:
I updated to just use the constant and gave it more helpful documentation.

Reply:
Nice, way clearer